### PR TITLE
Tensor mask changes

### PR DESCRIPTION
**Pull request by SundarRajan98**
Added changes for supporting tensor mask trainings

### CODE REVIEW

File: rocAL/include/api/rocal_api_data_transfer.h [hunk 1]
```diff
@@ -63,7 +63,7 @@ extern "C" RocalStatus ROCAL_API_CALL rocalToTensor(RocalContext rocal_context,
                                                     RocalTensorLayout tensor_format, RocalTensorOutputType tensor_output_type,
                                                     float multiplier0, float multiplier1, float multiplier2, float offset0,
                                                     float offset1, float offset2,
-                                                    bool reverse_channels, RocalOutputMemType output_mem_type);
+                                                    bool reverse_channels, RocalOutputMemType output_mem_type, int max_height = 0, int max_width = 0);
```
Comment (fgladwin) [inline rocAL/include/api/rocal_api_data_transfer.h:66]: max_width and height should be int?

File: rocAL/include/meta_data/meta_data_reader.h [hunk 2]
```diff
@@ -56,26 +56,35 @@ struct MetaDataConfig {
...
     unsigned out_img_height() const { return _out_img_height; }
     void set_out_img_width(unsigned out_img_width) { _out_img_width = out_img_width; }
     void set_out_img_height(unsigned out_img_height) { _out_img_height = out_img_height; }
+    void set_avoid_class_remapping(bool avoid_class_remapping) { _avoid_class_remapping = avoid_class_remapping; }
+    void set_aspect_ratio_grouping(bool aspect_ratio_grouping) { _aspect_ratio_grouping = aspect_ratio_grouping; }
 };
 
 class MetaDataReader {
+   private:
+    bool _aspect_ratio_grouping = false;
```
Comment (fgladwin) [inline rocAL/include/meta_data/meta_data_reader.h]: I don't think it should be done this way

File: rocAL/include/meta_data/meta_data_reader.h [hunk 3]
```diff
@@ -88,4 +97,7 @@ class MetaDataReader {
     virtual const std::map<std::string, std::shared_ptr<MetaData>>& get_map_content() = 0;
     virtual bool exists(const std::string& image_name) = 0;
     virtual bool set_timestamp_mode() = 0;
+    virtual ImgSize lookup_image_size(const std::string& image_name) { return {}; }
+    void set_aspect_ratio_grouping(bool aspect_ratio_grouping) { _aspect_ratio_grouping = aspect_ratio_grouping; }
+    bool aspect_ratio_grouping() const { return _aspect_ratio_grouping; }
```
Comment (fgladwin) [inline rocAL/include/meta_data/meta_data_reader.h]: This should also be like a virtual function only

File: rocAL/include/meta_data/meta_data_reader.h (hunk 2 above)
Comment (fgladwin) [reply rocAL/include/meta_data/meta_data_reader.h]: The COCO metadata reader only must have this if it is used there, Not in metadata reader
Please check

File: rocAL/rocAL_hip/rocal_hip_kernels.cpp [hunk 4]
```diff
@@ -124,25 +132,29 @@ Hip_CopyInt8ToNCHW_fp32(
...
         return;
     for (unsigned int n = 0; n < nchw.x; n++) {
         unsigned int srcIdx = (y * W + x) * C;
-        unsigned int dstIdx = y * W + x;
+        unsigned int dstIdx = y * maxOutW + x;
         // copy float3  pixels to dst
         const uchar *inp_img = &inp_image_u8[n * img_offset];
-        float *out_tensor = (float *)output_tensor + n * img_offset + dst_buf_offset;
+        float *out_tensor = (float *)output_tensor + n * out_img_offset + dst_buf_offset;
+        unsigned int stride = maxOutW * maxOutH;
```
Comment (fgladwin) [inline rocAL/rocAL_hip/rocal_hip_kernels.cpp]: Name it as cstride / channel stride something

File: rocAL/source/api/rocal_api_augmentation.cpp [hunk 5]
```diff
@@ -554,16 +554,20 @@ RocalTensor ROCAL_API_CALL
     try {
         if ((dest_width | dest_height | resize_longer | resize_shorter) == 0)
             THROW("Atleast one size 'dest_width' or 'dest_height' or 'resize_shorter' or 'resize_longer' must be specified")
-        if ((dest_width | dest_height) && (resize_longer | resize_shorter))
+        if ((dest_width | dest_height) && (resize_longer | resize_shorter) && (scaling_mode != RocalResizeScalingMode::ROCAL_SCALING_MODE_MIN_MAX))
             THROW("Only one method of specifying size can be used \ndest_width and/or dest_height\nresize_shorter\nresize_longer")
-        if (resize_longer && resize_shorter)
-            THROW("'resize_longer' and 'resize_shorter' cannot be passed together. They are mutually exclusive.")
+        if (resize_longer && resize_shorter && scaling_mode != RocalResizeScalingMode::ROCAL_SCALING_MODE_MIN_MAX)
+            THROW("'resize_longer' and 'resize_shorter' can only be passed together for min max scaling mode")
```
Comment (fgladwin) [inline rocAL/source/api/rocal_api_augmentation.cpp:562]: This statement is wrong

Either resize longer or resize shorter can be enabled

Both cannot be enabled at the same time, what are we inferring from here?

File: rocAL/source/api/rocal_api_augmentation.cpp [hunk 6]
```diff
@@ -554,16 +554,20 @@ RocalTensor ROCAL_API_CALL
...
 
         unsigned out_width, out_height;
         RocalResizeScalingMode resize_scaling_mode;
 
         // Change the scaling mode if resize_shorter or resize_longer is specified
-        if (resize_shorter) {
+        if (scaling_mode == RocalResizeScalingMode::ROCAL_SCALING_MODE_MIN_MAX) {
```
Comment (fgladwin) [inline rocAL/source/api/rocal_api_augmentation.cpp:568]: The else part is doing the same thing then why we need to add this?

File: rocAL/source/pipeline/master_graph.cpp [hunk 7]
```diff
@@ -556,11 +560,11 @@ MasterGraph::to_tensor(void *out_ptr, RocalTensorlayout format, float multiplier
             auto img_buffer = out_tensor;
             if (format == RocalTensorlayout::NHWC) {
                 HipExecCopyInt8ToNHWC(_device.resources()->hip_stream, (const void *)img_buffer, out_ptr, dest_buf_offset, n, c, h, w,
-                                      multiplier0, multiplier1, multiplier2, offset0, offset1, offset2, reverse_channels, fp16);
+                                      multiplier0, multiplier1, multiplier2, offset0, offset1, offset2, reverse_channels, fp16, max_height, max_width);
 
             } else {
                 HipExecCopyInt8ToNCHW(_device.resources()->hip_stream, (const void *)img_buffer, out_ptr, dest_buf_offset, n, c, h, w,
-                                      multiplier0, multiplier1, multiplier2, offset0, offset1, offset2, reverse_channels, fp16);
+                                      multiplier0, multiplier1, multiplier2, offset0, offset1, offset2, reverse_channels, fp16, max_height, max_width);
```
Comment (fgladwin) [inline rocAL/source/pipeline/master_graph.cpp:567]: The max_height and max_width which is an int in this API to_tensor is passed as unsigned into to this API

Please retain it as int or uint everywhere

File: rocAL/source/api/rocal_api_meta_data.cpp [hunk 8]
```diff
@@ -396,10 +421,8 @@ void
     }
 }
 
-void
-    ROCAL_API_CALL
-    rocalBoxEncoder(RocalContext p_context, std::vector<float>& anchors, float criteria,
-                    std::vector<float>& means, std::vector<float>& stds, bool offset, float scale) {
+void ROCAL_API_CALL rocalBoxEncoder(RocalContext p_context, std::vector<float>& anchors, float criteria,
```
Comment (fgladwin) [inline rocAL/source/api/rocal_api_meta_data.cpp]: Revert this change

File: rocAL/source/meta_data/coco_meta_data_reader.cpp [hunk 9]
```diff
@@ -178,10 +188,13 @@ void COCOMetaDataReader::read_all(const std::string &path) {
                         img_size.h = parser.GetInt();
                     } else if (0 == std::strcmp(internal_key, "file_name")) {
                         image_name = parser.GetString();
+                    } else if (0 == std::strcmp(internal_key, "id")) {
+                        image_id = parser.GetInt();
                     } else {
                         parser.SkipValue();
                     }
                 }
+                _map_img_names.insert(pair<int, std::string>(image_id, image_name));
```
Comment (fgladwin) [inline rocAL/source/meta_data/coco_meta_data_reader.cpp]: Can we rename this as _map_image_names_to_id so it will be evident?

File: rocAL/source/meta_data/meta_node_resize_mirror_normalize.cpp [hunk 10]
```diff
@@ -75,6 +75,11 @@ void ResizeMirrorNormalizeMetaNode::update_parameters(pMetaDataBatch input_meta_
             bb_coords.push_back(coords_buf[j]);
             bb_labels.push_back(labels_buf[j]);
         }
+        // get roi width and height of output image
+        auto img_roi_size = input_meta_data->get_img_roi_sizes_batch()[i];
```
Comment (fgladwin) [inline rocAL/source/meta_data/meta_node_resize_mirror_normalize.cpp:79]: Can we not update directly using reference or something

We get value update and then set, Try obtaining reference then update directly. Check if that would work

File: rocAL/include/meta_data/meta_data_reader.h [hunk 11]
```diff
@@ -56,26 +56,35 @@ struct MetaDataConfig {
     unsigned _frame_stride;
     unsigned _out_img_width;
     unsigned _out_img_height;
+    bool _avoid_class_remapping;
+    bool _aspect_ratio_grouping;
 
    public:
-    MetaDataConfig(const MetaDataType& type, const MetaDataReaderType& reader_type, const std::string& path, const std::map<std::string, std::string>& feature_key_map = std::map<std::string, std::string>(), const std::string file_prefix = std::string(), const unsigned& sequence_length = 3, const unsigned& frame_step = 3, const unsigned& frame_stride = 1)
-        : _type(type), _reader_type(reader_type), _path(path), _feature_key_map(feature_key_map), _file_prefix(file_prefix), _sequence_length(sequence_length), _frame_step(frame_step), _frame_stride(frame_stride) {}
+    MetaDataConfig(const MetaDataType& type, const MetaDataReaderType& reader_type, const std::string& path, const std::map<std::string, std::string>& feature_key_map = std::map<std::string, std::string>(), const std::string file_prefix = std::string(), const unsigned& sequence_length = 3, const unsigned& frame_step = 3, const unsigned& frame_stride = 1, bool avoid_class_remapping = false)
```
Comment (fgladwin) [inline rocAL/include/meta_data/meta_data_reader.h]: This avoid class remapping does it need to be part of init?

Anyways we have API to set it. Please check

File: rocAL/source/pipeline/ring_buffer.cpp [hunk 12]
```diff
@@ -22,8 +22,6 @@ THE SOFTWARE.
 
 #include "ring_buffer.h"
 
-#include <device_manager.h>
```
Comment (fgladwin) [inline rocAL/source/pipeline/ring_buffer.cpp]: This needs removal? The develop branch has this added Please check

File: rocAL/include/api/rocal_api_meta_data.h [hunk 13]
```diff
@@ -209,6 +214,13 @@ extern "C" RocalTensorList ROCAL_API_CALL rocalGetBoundingBoxCords(RocalContext
  */
 extern "C" void ROCAL_API_CALL rocalGetImageSizes(RocalContext rocal_context, int* buf);
 
+/*! \brief get ROI image sizes
+ * \ingroup group_rocal_meta_data
+ * \param [in] rocal_context rocal context
+ * \param [out] buf The user's buffer that will be filled with ROI image size info for the images in the output batch
+ */
+extern "C" void ROCAL_API_CALL rocalGetROIImageSizes(RocalContext rocal_context, int* buf);
```
Comment (SundarRajan98) [inline rocAL/include/api/rocal_api_meta_data.h:222]: What is the difference between rocalGetImageSizes and rocalGetROIImageSizes?

File: rocAL/include/meta_data/meta_data_reader.h (hunk 2 above)
Comment (SundarRajan98) [inline rocAL/include/meta_data/meta_data_reader.h]: Both in MetaDataReader and MetaDataConfig the _aspect_ratio_grouping variable is required?

File: rocAL/rocAL_hip/rocal_hip_kernels.cpp (hunk 4 above)
Comment (SundarRajan98) [inline rocAL/rocAL_hip/rocal_hip_kernels.cpp]: stride is a constant can be moved outside the loop, so that it won't be computed every time

File: rocAL/source/readers/image/coco_file_source_reader.cpp [hunk 14]
```diff
@@ -85,9 +85,53 @@ Reader::Status COCOFileSourceReader::initialize(ReaderConfig desc) {
...
+
+    if (_meta_data_reader && _meta_data_reader->aspect_ratio_grouping()) {
+        // calculate the aspect ratio for each file and create a pair of <filename, aspect_ratio>
+        std::vector<std::pair<std::string, float>> file_aspect_ratio_pair(_file_names.size());
+        for (size_t i = 0; i < _file_names.size(); i++) {
+            auto filename = _file_names[i];
+            std::string base_filename = filename.substr(filename.find_last_of("/\\") + 1);
+            auto img_size = _meta_data_reader->lookup_image_size(base_filename);
+            auto aspect_ratio = static_cast<float>(img_size.h) / img_size.w;
+            file_aspect_ratio_pair[i] = std::make_pair(filename, aspect_ratio);
```
Comment (fgladwin) [inline rocAL/source/readers/image/coco_file_source_reader.cpp:97]: Can we sort while inserting itself?

File: rocAL/rocAL_hip/rocal_hip_kernels.cpp [hunk 15]
```diff
@@ -204,23 +221,30 @@ int HipExecCopyInt8ToNHWC(
...
-    unsigned int fp16) {
+    unsigned int fp16,
+    const unsigned max_output_height,
+    const unsigned max_output_width) {
     int localThreads_x = 16, localThreads_y = 16;
+    uint2 outDims;
+    if ((max_output_height == 0) || (max_output_width == 0))
+        outDims = make_uint2(h, w);
+    else
+        outDims = make_uint2(max_output_height, max_output_width);
```
Comment (SundarRajan98) [inline rocAL/rocAL_hip/rocal_hip_kernels.cpp]: Check if we can use std::max instead of if conditions here

File: rocAL/rocAL_hip/rocal_hip_kernels.cpp [hunk 16]
```diff
@@ -204,23 +221,30 @@ int HipExecCopyInt8ToNHWC(
...
     int globalThreads_x = w, globalThreads_y = h;
     if (!fp16) {
         hipLaunchKernelGGL(Hip_CopyInt8ToNHWC_fp32,
                            dim3(ceil((float)globalThreads_x / localThreads_x), ceil((float)globalThreads_y / localThreads_y)),
                            dim3(localThreads_x, localThreads_y),
                            0, stream, (const uchar *)inp_image_u8, output_tensor, dst_buf_offset,
-                           make_uint4(n, c, h, w),
+                           make_uint4(n, c, h, w), outDims,
                            make_float3(multiplier0, multiplier1, multiplier2), make_float3(offset0, offset1, offset2),
```
Comment (SundarRajan98) [inline rocAL/rocAL_hip/rocal_hip_kernels.cpp:240]: Move outDims to new line

File: rocAL/source/readers/image/coco_file_source_reader.cpp [hunk 17]
```diff
@@ -173,8 +217,27 @@ int COCOFileSourceReader::release() {
...
 
 void COCOFileSourceReader::reset() {
-    if (_shuffle)
+    if (_meta_data_reader && _meta_data_reader->aspect_ratio_grouping()) {
+        _file_names = _sorted_file_names;
+        // Calculate the mid element which divides the aspect ratios into two groups (<=1.0 and >1.0)
+        auto mid = std::upper_bound(_aspect_ratios.begin(), _aspect_ratios.end(), 1.0f) - _aspect_ratios.begin();
+        if (_shuffle) {
+            // Shuffle within groups using the mid element as the limit - [start, mid) and [mid, last)
+            std::random_shuffle(_file_names.begin(), _file_names.begin() + mid);
```
Comment (fgladwin) [inline rocAL/source/readers/image/coco_file_source_reader.cpp]: If this block is repetitive try to modularize

File: rocAL_pybind/amd/rocal/pipeline.py [hunk 18]
```diff
@@ -146,6 +146,11 @@ def define_graph(self):
     def get_handle(self):
         return self._handle
 
+    def copyToExternalTensor(self, array,  multiplier, offset, reverse_channels, tensor_format, tensor_dtype, max_height=0, max_width=0):
+
```
Comment (fgladwin) [inline rocAL_pybind/amd/rocal/pipeline.py]: why this empty line here?

File: rocAL_pybind/amd/rocal/pipeline.py [hunk 19]
```diff
@@ -146,6 +146,11 @@ def define_graph(self):
...
+        b.rocalToTensor(self._handle, ctypes.c_void_p(array.data_ptr()), tensor_format, tensor_dtype,
+                                    multiplier[0], multiplier[1], multiplier[2], offset[0], offset[1], offset[2], (1 if reverse_channels else 0), self._output_memory_type, max_height, max_width)
```
Comment (fgladwin) [inline rocAL_pybind/amd/rocal/pipeline.py]: Format this

File: rocAL_pybind/rocal_pybind.cpp [hunk 20]
```diff
@@ -450,6 +469,41 @@ PYBIND11_MODULE(rocal_pybind, m) {
...
+        py::list complete_list;
+        int poly_cnt = 0;
+        int prev_object_cnt = 0;
+        auto mask_count_buf = mask_count.request();
+        int *mask_count_ptr = static_cast<int *>(mask_count_buf.ptr);
+        for (int i = 0; i < bbox_labels->size(); i++) {  // nbatchSize
+            float *mask_buffer = static_cast<float *>(mask_data->at(i)->buffer());
+            py::list poly_batch_list;
+            for (unsigned j = prev_object_cnt; j < bbox_labels->at(i)->dims().at(0) + prev_object_cnt; j++) {
+                py::list single_image;
```
Comment (fgladwin) [inline rocAL_pybind/rocal_pybind.cpp]: Why is this called single_image change name

File: rocAL/source/api/rocal_api_augmentation.cpp [hunk 21]
```diff
@@ -609,6 +613,10 @@ RocalTensor ROCAL_API_CALL
                 max_out_height = maximum_size[1] ? maximum_size[1] : max_out_height;
             }
         }
+        if (scaling_mode == RocalResizeScalingMode::ROCAL_SCALING_MODE_MIN_MAX) {
```
Comment (SundarRajan98) [inline rocAL/source/api/rocal_api_augmentation.cpp:618]: Try to combine if statements for RocalResizeScalingMode::ROCAL_SCALING_MODE_MIN_MAX) at L616, L566 into single case if possible

File: rocAL/source/pipeline/ring_buffer.cpp (hunk 12 above)
Comment (SundarRajan98) [inline rocAL/source/pipeline/ring_buffer.cpp]: Why is this removed?

File: rocAL_pybind/rocal_pybind.cpp [hunk 22]
```diff
@@ -84,6 +84,18 @@ py::object wrapper_image_name(RocalContext context, int array_len) {
     return py::bytes(s);
 }
 
+py::object wrapper_copy_to_tensor(RocalContext context, py::object p,
+                                  RocalTensorLayout tensor_format, RocalTensorOutputType tensor_output_type, float multiplier0,
+                                  float multiplier1, float multiplier2, float offset0, float offset1, float offset2,
+                                  bool reverse_channels, RocalOutputMemType output_mem_type, int max_height, int max_width) {
```
Comment (SundarRajan98) [inline rocAL_pybind/rocal_pybind.cpp]: Arguments to this functions looks not aligned
Please fix the alignment

File: rocAL_pybind/rocal_pybind.cpp [hunk 23]
```diff
@@ -84,6 +84,18 @@ py::object wrapper_image_name(RocalContext context, int array_len) {
...
+    auto ptr = ctypes_void_ptr(p);
+    // call pure C++ function
+    int status = rocalToTensor(context, ptr, tensor_format, tensor_output_type, multiplier0,
+                               multiplier1, multiplier2, offset0, offset1, offset2,
```
Comment (SundarRajan98) [inline rocAL_pybind/rocal_pybind.cpp:94]: Same comment
Fix the alignment here

File: rocAL_pybind/amd/rocal/types.py [hunk 24]
```diff
@@ -79,6 +79,7 @@
 from rocal_pybind.types import SCALING_MODE_STRETCH
 from rocal_pybind.types import SCALING_MODE_NOT_SMALLER
 from rocal_pybind.types import SCALING_MODE_NOT_LARGER
+from rocal_pybind.types import SCALING_MODE_MIN_MAX
```
Comment (SundarRajan98) [inline rocAL_pybind/amd/rocal/types.py:82]: from rocal_pybind.types multiple modes are imported
You can condense all of these in single line

File: rocAL/source/readers/image/coco_file_source_reader.cpp [hunk 25]
```diff
@@ -85,9 +85,53 @@ Reader::Status COCOFileSourceReader::initialize(ReaderConfig desc) {
...
+            _aspect_ratios.push_back(aspect_ratio);
+        };
+
+        // sort the <filename, aspect_ratio> pairs according to aspect ratios
+        std::sort(file_aspect_ratio_pair.begin(), file_aspect_ratio_pair.end(), [](auto &lop, auto &rop) { return lop.second < rop.second; });
+
+        // extract sorted file_names
+        std::transform(file_aspect_ratio_pair.begin(), file_aspect_ratio_pair.end(), std::back_inserter(_sorted_file_names), [](auto &pair) { return pair.first; });
+        // extract sorted aspect ratios
+        _aspect_ratios.clear();
```
Comment (SundarRajan98) [inline rocAL/source/readers/image/coco_file_source_reader.cpp:107]: Why are we clearing the values in _aspect_ratios vector?

**Review by SundarRajan98 — COMMENTED** [review]
Please address review comments

File: rocAL/include/meta_data/coco_meta_data_reader.h [hunk 26]
```diff
@@ -32,25 +32,27 @@ class COCOMetaDataReader : public MetaDataReader {
    public:
     void init(const MetaDataConfig& cfg, pMetaDataBatch meta_data_batch) override;
     void lookup(const std::vector<std::string>& image_names) override;
+    ImgSize lookup_image_size(const std::string& image_name) override;
```
Comment (shobiloganat) [inline rocAL/include/meta_data/coco_meta_data_reader.h:35]: image_name -> image_names

File: rocAL/include/meta_data/meta_data_reader.h (hunk 2 above)
Comment (SundarRajan98) [reply rocAL/include/meta_data/meta_data_reader.h]: This is declared here becoz its needed in coco_file_source_reader which only has access to MetadataReader definition. So it cannot access an internal variable of coco metadata reader class.

File: rocAL/source/api/rocal_api_augmentation.cpp (hunk 5 above)
Comment (SundarRajan98) [reply rocAL/source/api/rocal_api_augmentation.cpp:562]: We are passing resize_min and resize_max values as resize_shorter and resize_longer values to RMN node. So that's why we are allowing to enable both of them for this particular scaling mode

File: rocAL/source/api/rocal_api_augmentation.cpp (hunk 6 above)
Comment (SundarRajan98) [reply rocAL/source/api/rocal_api_augmentation.cpp:568]: It won't get to the else condition since resize_shorter and resize_longer is set in this scaling mode so I had to create a new if condition for this case

File: rocAL/source/meta_data/meta_node_resize_mirror_normalize.cpp (hunk 10 above)
Comment (SundarRajan98) [reply rocAL/source/meta_data/meta_node_resize_mirror_normalize.cpp:79]: I'm not updating it in the same metadata ptr. It was the same ptr before so it worked but now I'm getting the roi_size from input_meta_data and updating it in output_meta_data

File: rocAL/source/pipeline/ring_buffer.cpp (hunk 12 above)
Comment (SundarRajan98) [reply rocAL/source/pipeline/ring_buffer.cpp]: I removed it since it was also included in ring buffer header file. Let me know if this is needed or not

File: rocAL/include/api/rocal_api_meta_data.h (hunk 13 above)
Comment (SundarRajan98) [reply rocAL/include/api/rocal_api_meta_data.h:222]: rocalGetImageSizes gives the size of the final output buffer. rocalGetROIImageSizes gives the ROI of the final output buffer

File: rocAL/include/meta_data/meta_data_reader.h (hunk 2 above)
Comment (SundarRajan98) [reply rocAL/include/meta_data/meta_data_reader.h]: We get the aspect ratio grouping parameter in master graph where we can only set metadata config params. Only in coco metadata reader init, we are fetching these values from metadata config and updating it in metadata reader

File: rocAL/rocAL_hip/rocal_hip_kernels.cpp (hunk 15 above)
Comment (SundarRajan98) [reply rocAL/rocAL_hip/rocal_hip_kernels.cpp]: h/w will always be non-zero and mostly be larger than the ROI width and height passed by user so using std::max will only result in bigger buffer normalization and copy

File: rocAL/source/api/rocal_api_augmentation.cpp (hunk 21 above)
Comment (SundarRajan98) [reply rocAL/source/api/rocal_api_augmentation.cpp:618]: Both of the if statements are added near similar if statements so grouping them together will be confusing

File: rocAL_pybind/rocal_pybind.cpp (hunk 22 above)
Comment (SundarRajan98) [reply rocAL_pybind/rocal_pybind.cpp]: I used C++ formatter in vscode for this file and could not find any formatting issues here

File: rocAL_pybind/rocal_pybind.cpp (hunk 23 above)
Comment (SundarRajan98) [reply rocAL_pybind/rocal_pybind.cpp:94]: I used C++ formatter in vscode for this file and could not find any formatting issues here

File: rocAL_pybind/amd/rocal/types.py (hunk 24 above)
Comment (SundarRajan98) [reply rocAL_pybind/amd/rocal/types.py:82]: I tried the same in previous pybind PR. Lakshmi stated that it throws warnings so I reverted back to this

File: rocAL/source/readers/image/coco_file_source_reader.cpp (hunk 25 above)
Comment (SundarRajan98) [reply rocAL/source/readers/image/coco_file_source_reader.cpp:107]: I am sorting the aspect ratios for the file_names and storing it in the same vector so first I'm clearing it and using back_inserter to insert them as new_elements and not updating them in place

File: rocAL/include/meta_data/coco_meta_data_reader.h (hunk 26 above)
Comment (shobiloganat) [reply rocAL/include/meta_data/coco_meta_data_reader.h:35]: What naming convention does ImgSize follows?

File: rocAL/include/meta_data/coco_meta_data_reader.h [hunk 27]
```diff
@@ -32,25 +32,29 @@ class COCOMetaDataReader : public MetaDataReader {
...
     void read_all(const std::string& path) override;
     void release(std::string image_name);
     void release() override;
     void print_map_contents();
     bool set_timestamp_mode() override { return false; }
-
     const std::map<std::string, std::shared_ptr<MetaData>>& get_map_content() override { return _map_content; }
+    void set_aspect_ratio_grouping(bool aspect_ratio_grouping) override { _aspect_ratio_grouping = aspect_ratio_grouping; }
+    bool aspect_ratio_grouping() const override { return _aspect_ratio_grouping; }
```
Comment (shobiloganat) [inline rocAL/include/meta_data/coco_meta_data_reader.h]: Change this to aspect_ratio_grouping() -> get_aspect_ratio_grouping()

File: rocAL/include/meta_data/coco_meta_data_reader.h [hunk 28]
```diff
@@ -32,25 +32,29 @@ class COCOMetaDataReader : public MetaDataReader {
...
     int meta_data_reader_type;
+    bool _avoid_class_remapping;
     void add(std::string image_name, BoundingBoxCords bbox, Labels labels, ImgSize image_size, int image_id = 0);
-    void add(std::string image_name, BoundingBoxCords bbox, Labels labels, ImgSize image_size, MaskCords mask_cords, std::vector<int> polygon_count, std::vector<std::vector<int>> vertices_count);  // To add Mask coordinates to Metadata struct
+    void add(std::string image_name, BoundingBoxCords bbox, Labels labels, ImgSize image_size, MaskCords mask_cords, std::vector<int> polygon_count, std::vector<std::vector<int>> vertices_count, int image_id = 0);  // To add Mask coordinates to Metadata struct
     bool exists(const std::string& image_name) override;
     std::map<std::string, std::shared_ptr<MetaData>> _map_content;
     std::map<std::string, std::shared_ptr<MetaData>>::iterator _itr;
     std::map<std::string, ImgSize> _map_img_sizes;
+    std::map<int, std::string> _map_img_names;
```
Comment (shobiloganat) [inline rocAL/include/meta_data/coco_meta_data_reader.h]: Add a comment about what's the int in this map img names

File: rocAL/include/meta_data/meta_data_reader.h [hunk 29]
```diff
@@ -66,16 +68,23 @@ struct MetaDataConfig {
...
+    bool aspect_ratio_grouping() const { return _aspect_ratio_grouping; }
     unsigned sequence_length() const { return _sequence_length; }
     unsigned frame_step() const { return _frame_step; }
     unsigned frame_stride() const { return _frame_stride; }
     unsigned out_img_width() const { return _out_img_width; }
     unsigned out_img_height() const { return _out_img_height; }
     void set_out_img_width(unsigned out_img_width) { _out_img_width = out_img_width; }
     void set_out_img_height(unsigned out_img_height) { _out_img_height = out_img_height; }
+    void set_avoid_class_remapping(bool avoid_class_remapping) { _avoid_class_remapping = avoid_class_remapping; }
+    void set_aspect_ratio_grouping(bool aspect_ratio_grouping) { _aspect_ratio_grouping = aspect_ratio_grouping; }
```
Comment (shobiloganat) [inline rocAL/include/meta_data/meta_data_reader.h:81]: Is this required in both meta data reader and coco meta data reader?

File: rocAL/rocAL_hip/rocal_hip_kernels.cpp [hunk 30]
```diff
@@ -40,19 +41,21 @@ Hip_CopyInt8ToNHWC_fp32(
...
         unsigned int srcIdx = (y * W + x) * C;  // src is RGB
-        unsigned int dstIdx = (y * W + x) * C;
+        unsigned int dstIdx = (y * maxOutW + x) * C;
         // copy float3  pixels to dst
         if (C == 3) {
             float3 dst;
-
             const uchar *inp_img = &inp_image_u8[n * img_offset];
-            float *out_tensor = (float *)((float *)output_tensor + dst_buf_offset + n * img_offset);
+            float *out_tensor = (float *)((float *)output_tensor + dst_buf_offset + n * out_img_offset);
```
Comment (shobiloganat) [inline rocAL/rocAL_hip/rocal_hip_kernels.cpp]: Why some of the variables are in camel case and some in snake case?

File: rocAL/source/api/rocal_api_augmentation.cpp (hunk 5 above)
Comment (shobiloganat) [reply rocAL/source/api/rocal_api_augmentation.cpp:562]: Add that in a comment to explain here

File: rocAL/source/meta_data/coco_meta_data_reader.cpp (hunk 9 above)
Comment (shobiloganat) [reply rocAL/source/meta_data/coco_meta_data_reader.cpp]: Please check this.

File: rocAL/source/readers/image/coco_file_source_reader.cpp [hunk 31]
```diff
@@ -172,9 +203,30 @@ int COCOFileSourceReader::release() {
     return 0;
 }
 
+void COCOFileSourceReader::shuffle_with_aspect_ratios() {
+    // Calculate the mid element which divides the aspect ratios into two groups (<=1.0 and >1.0)
+    auto mid = std::upper_bound(_aspect_ratios.begin(), _aspect_ratios.end(), 1.0f) - _aspect_ratios.begin();
+    // Shuffle within groups using the mid element as the limit - [start, mid) and [mid, last)
+    std::random_shuffle(_file_names.begin(), _file_names.begin() + mid);
+    std::random_shuffle(_file_names.begin() + mid, _file_names.end());
```
Comment (shobiloganat) [inline rocAL/source/readers/image/coco_file_source_reader.cpp:211]: Can we not combine this in a single line by nesting?

File: rocAL_pybind/rocal_pybind.cpp [hunk 32]
```diff
@@ -450,6 +469,41 @@ PYBIND11_MODULE(rocal_pybind, m) {
...
+        int *polygon_size_ptr = static_cast<int *>(buf.ptr);
+        // call pure C++ function
+        rocalTensorList *mask_data = rocalGetMaskCoordinates(context, polygon_size_ptr);
+        rocalTensorList *bbox_labels = rocalGetBoundingBoxLabel(context);
+        py::list complete_list;
+        int poly_cnt = 0;
+        int prev_object_cnt = 0;
+        auto mask_count_buf = mask_count.request();
+        int *mask_count_ptr = static_cast<int *>(mask_count_buf.ptr);
+        for (int i = 0; i < bbox_labels->size(); i++) {  // nbatchSize
```
Comment (shobiloganat) [inline rocAL_pybind/rocal_pybind.cpp]: Add a comment about what's done here.

File: rocAL/include/meta_data/meta_data_reader.h [hunk 33]
```diff
@@ -88,4 +97,7 @@ class MetaDataReader {
     virtual const std::map<std::string, std::shared_ptr<MetaData>>& get_map_content() = 0;
     virtual bool exists(const std::string& image_name) = 0;
     virtual bool set_timestamp_mode() = 0;
+    virtual ImgSize lookup_image_size(const std::string& image_name) { return {}; }
+    virtual void set_aspect_ratio_grouping(bool aspect_ratio_grouping) { return; }
```
Comment (shobiloganat) [inline rocAL/include/meta_data/meta_data_reader.h:101]: Is the set aspect ratio grouping required in both meta data reader and config?
Please check.

**Review by shobiloganat — COMMENTED** [review]
Please address the comments.

File: rocAL/source/pipeline/master_graph.cpp [hunk 34]
```diff
@@ -669,34 +675,37 @@ MasterGraph::to_tensor(void *out_ptr, RocalTensorlayout format, float multiplier
...
+                                        __m256i pix0 = _mm256_loadu_si256((const __m256i *)in_buffer_row);
+                                        pix0 = _mm256_permutevar8x32_epi32(pix0, _mm256_setr_epi32(0, 1, 2, 3, 3, 4, 5, 6));
+                                        fB = _mm256_cvtepi32_ps(_mm256_shuffle_epi8(pix0, mask_R));
+                                        fG = _mm256_cvtepi32_ps(_mm256_shuffle_epi8(pix0, mask_G));
+                                        fR = _mm256_cvtepi32_ps(_mm256_shuffle_epi8(pix0, mask_B));
+                                        fB = _mm256_mul_ps(fB, pmul0);
+                                        fG = _mm256_mul_ps(fG, pmul1);
+                                        fR = _mm256_mul_ps(fR, pmul2);
+                                        fB = _mm256_add_ps(fB, padd0);
+                                        fG = _mm256_add_ps(fG, padd1);
```
Comment (sampath-amd) [inline rocAL/source/pipeline/master_graph.cpp]: Please use fmadd  instead of add and multiply as separate instructions

File: rocAL/include/meta_data/coco_meta_data_reader.h (hunk 26 above)
Comment (SundarRajan98) [reply rocAL/include/meta_data/coco_meta_data_reader.h:35]: Only a single image name is passed so named it image_name. ImgSize is defined in metadata.h where all structs are following PascalCase

File: rocAL/include/meta_data/meta_data_reader.h (hunk 29 above)
Comment (SundarRajan98) [reply rocAL/include/meta_data/meta_data_reader.h:81]: In master_graph.cpp, where we create the coco_reader, we only have access to the MetaDataConfig so we set aspect_ratio_grouping there. When coco metadata reader is initialized, it uses its own aspect_ratio_grouping function to set its inherited variable. This is needed becoz we do aspect ratio grouping in coco file source reader which has access to the coco metadata reader but not its metadata config

File: rocAL/include/meta_data/meta_data_reader.h (hunk 33 above)
Comment (SundarRajan98) [reply rocAL/include/meta_data/meta_data_reader.h:101]: In master_graph.cpp, where we create the coco_reader, we only have access to the MetaDataConfig so we set aspect_ratio_grouping there. When coco metadata reader is initialized, it uses its own aspect_ratio_grouping function to set its inherited variable. This is needed becoz we do aspect ratio grouping in coco file source reader which has access to the coco metadata reader but not its metadata config
